presto-main/src/main/java/io/prestosql/operator/window/WindowPartition.java

```diff
@@ -546,4 +605,201 @@ private long getFrameValue(int channel, String type)
         checkCondition(value >= 0, INVALID_WINDOW_FRAME, "Window frame %s offset must not be negative", value);
         return value;
     }
+
+    private GroupsFrame getFrameRange(FrameInfo frameInfo, GroupsFrame recentFrame)
+    {
+        Type startType = frameInfo.getStartType();
+        Type endType = frameInfo.getEndType();
+        int start;
+        int end;
+
+        // Frame is unbounded or defined by current peer group
+        if ((startType == UNBOUNDED_PRECEDING || startType == CURRENT_ROW) &&
+                (endType == UNBOUNDED_FOLLOWING || endType == CURRENT_ROW)) {
+            start = startType == UNBOUNDED_PRECEDING ? 0 : peerGroupStart - partitionStart;
+            end = endType == UNBOUNDED_FOLLOWING ? partitionEnd - partitionStart - 1 : peerGroupEnd - partitionStart - 1;
+            return new GroupsFrame(start, GroupsFrame.ignoreIndex(), end, GroupsFrame.ignoreIndex());
+        }
+
+        // Frame definition has at least one of: X PRECEDING, Y FOLLOWING
+        int startGroupIndex;
+        int endGroupIndex;
+        if (startType == UNBOUNDED_PRECEDING) {
+            start = 0;
+            startGroupIndex = GroupsFrame.ignoreIndex();
+        }
+        else if (startType == CURRENT_ROW) {
+            start = peerGroupStart - partitionStart;
+            startGroupIndex = GroupsFrame.ignoreIndex();
+        }
+        else if (startType == PRECEDING) {
+            PositionAndGroup frameStart = seek(toIntExact(currentGroupIndex - getStartValue(frameInfo)), recentFrame.getStart(), recentFrame.getStartGroupIndex(), seekGroupStart, () -> new PositionAndGroup(0, 0));
```

Comment:
Why is the last argument passed as a lambda? It doesn't take any arguments, so there's no much point in computing it lazily.

Reply:
It is a lambda, because it uses `lastPeerGroup`, which is variable.
I could pass `lastPeerGroup` as an argument to the lambda.

Reply:
Yes, that's much clearer.

Reply:
Cool. It's already done this way. Ready for review.